themes/default/content/blog/troubleshooting-builds-with-diagnostics/index.md

```diff
@@ -0,0 +1,154 @@
+
+```bash
+pulumi up -v=5 --logflow
+```
+
+In practice, we generally use all of these together because that way, you don't need to know which type of provider you might be calling:
+
+```bash
+TF_LOG=DEBUG pulumi up -v=9 --logflow
+```
+
+If you really, really want to go down the rabbit hole, there's also an environment variable to help debug promise leaks. If you have a promise leak somewhere in your code, Pulumi [prints an error message](https://github.com/pulumi/pulumi/blob/f8a9698ca9599240223bed8305bbdfe4c4a5446a/sdk/nodejs/runtime/debuggable.ts#L36-L46) that tells you to use the `PULUMI_DEBUG_PROMISE_LEAKS` environment variable to rerun your command:
+
+```go
+export function leakedPromises(): [Set<Promise<any>>, string] {
+    const leaked = leakCandidates;
+    const promisePlural = leaked.size === 0 ? "promise was" : "promises were";
+    const message = leaked.size === 0 ? "" :
+        `The Pulumi runtime detected that ${leaked.size} ${promisePlural} still active\n` +
+        "at the time that the process exited. There are a few ways that this can occur:\n" +
+        "  * Not using `await` or `.then` on a Promise returned from a Pulumi API\n" +
+        "  * Introducing a cyclic dependency between two Pulumi Resources\n" +
+        "  * A bug in the Pulumi Runtime\n" +
+        "\n" +
+        "Leaving promises active is probably not what you want. If you are unsure about\n" +
+        "why you are seeing this message, re-run your program "
+            + "with the `PULUMI_DEBUG_PROMISE_LEAKS`\n" +
+        "environment variable. The Pulumi runtime will then print out additional\n" +
+        "debug information about the leaked promises.";
+
+    if (debugPromiseLeaks) {
+        for (const leak of leaked) {
+            console.error("Promise leak detected:");
+            console.error(promiseDebugString(leak));
+        }
+    }
+
+    leakCandidates = new Set<Promise<any>>();
+    return [leaked, message];
+}
+```
+
+The variable causes Pulumi to dump the context, stack trace, and error for the leak. It's very rare you'll need to do this kind of debugging, but in case you do:
+
+```bash
+PULUMI_DEBUG_PROMISE_LEAKS=true pulumi up -v=11
+```
+
+### Data return
+
+Those flags, though, don't tell the CLI where to send the data for you to examine. You might be working on a system that doesn't allow you to write logs anywhere, you might be streaming that data to another platform, or you might be able to write to a file to examine or parse later. If you want to write the diagnostic data to stdout/stderr, you need to add the `--logtostderr` flag:
+
+```bash
+TF_LOG=DEBUG pulumi up -v=9 --logflow --logtostderr
+```
+
+If you want to write that diagnostic data to a file and not to stdout/stderr, you can use your standard output pipe to point the data to a file (in this example, we're appending to any existing logfile):
```

Comment:
I’ve not yet seen a use case for Pulumi logs where you’d want to append.  This is rather different than service logging.